### DISCUSSION TITLE
Unable to get certificate renewal working through SCEP

### DISCUSSION THREAD

**Discussion by FelixG1316** [strongswan/strongswan#1689] [Q&A] [answered]
Hi everyone,

I have to support SCEP on a device to allow it to get certificates from a PKI and renew these certificates with SCEP.

I use strongswan as SCEP client and the [openxpki docker image](https://github.com/openxpki/openxpki-docker) as PKI. The PKI is just for test, so I used the `sampleconfig.sh` script which is provided with openxpki to install a test config. Then, I manage to enroll a certificate with SCEP.

My issue is when I want to renew this certificate. I followed the examples from the [strongSwan documentation](https://docs.strongswan.org/docs/5.9/pki/pkiScep.html#_examples), but I got a `400 Bad Request` reply from the server on the operation `PKIOperation`.

```
$ pki --scep --url http://openxpki_server:8080/scep --cacert ca.pem --cacert-sig ca-1.pem --cacert-enc ra.pem --in newscep.key --dn "C=FR, CN=my_hostname" --san "my_hostname" --cert scep.crt --key scep.key --outform pem
transaction ID: 48C611A24622D77FEF621C56AEC24E972DC9D176
did not receive a valid SCEP response: HTTP 400
```

On the server I can observe the following logs.

```
2023/05/09 15:03:29 ERR Unable to parse PKCS10: decode: decode error 06<=>30 4 8 certificationRequestInfo at /usr/share/perl5/Convert/ASN1/_decode.pm line 113.
Cannot handle input or missing ASN.1 definitions at /usr/share/perl5/Crypt/PKCS10.pm line 756.
    Crypt::PKCS10::_new(undef, undef, undef, "ignoreNonBase64", 1, "verifySignature", 1) called at /usr/share/perl5/Crypt/PKCS10.pm line 607
    eval {...} called at /usr/share/perl5/Crypt/PKCS10.pm line 604
    Crypt::PKCS10::new("Crypt::PKCS10", "0\x{82}\x{d}\x{bb}\x{6}\x{9}*\x{86}H\x{86}\x{f7}\x{d}\x{1}\x{7}\x{2}\x{a0}\x{82}\x{d}\x{ac}0\x{82}\x{d}\x{a8}\x{2}\x{1}\x{1}1\x{f}0\x{d}\x{6}\x{9}`\x{86}H\x{1}e\x{3}\x{4}\x{2}\x{1}\x{5}\x{0}0\x{82}\x{5}!\x{6}\x{9}*\x{86}H\x{86}\x{f7}\x{d}\x{1}\x{7}\x{1}\x{a0}\x{82}\x{5}"..., "ignoreNonBase64", 1, "verifySignature", 1) called at /usr/share/perl5/OpenXPKI/Client/Service/Base.pm line 188
    OpenXPKI::Client::Service::Base::handle_enrollment_request(OpenXPKI::Client::Service::SCEP=HASH(0x55e1ad5c6d80), CGI::Fast=HASH(0x55e1ad5f2060)) called at /usr/lib/cgi-bin/scepv3.fcgi line 101
 [pid=78|ep=]
2023/05/09 15:03:29 WAR Client error / malformed request badRequest [pid=78|ep=]
2023/05/09 15:03:29 INF Disconnect client [pid=78|ep=]
```

I read, still from the [strongSwan documentation](https://docs.strongswan.org/docs/5.9/pki/pki.html#_configuration), that:
> Some SCEP servers (e.g. [openxpki](https://github.com/openxpki/openxpki)) are incorrectly doing certificate renewal via [msgType](https://datatracker.ietf.org/doc/html/rfc8894#section-3.2.1.2) PKCSReq (19) instead of RenewalReq (17)

I understood there is a known issue with openxpki, and we can use the `scep.renewal_via_pkcs_req` option to deal with it. I found this option in `/etc/strongswan.d/pki.conf`, I tried to set it to `yes`, then to `no`, but it did not change anything.

I guess I misunderstood something, can someone help me to understand please ? 🙏

**Comment by tobiasbrunner** [comment]
> I found this option in `/etc/strongswan.d/pki.conf`, I tried to set it to `yes`, then to `no`, but it did not change anything.

You'll definitely have to set it to `yes`. Can you please share the contents of that file. Also, does your `/etc/strongswan.conf` file include that config snippet? (e.g. via an `include strongswan.d/*.conf` statement at the end, like the [default file](https://docs.strongswan.org/docs/5.9/config/strongswanDir.html))

**Reply by FelixG1316** [reply]
Hi @tobiasbrunner, yes the `strongswan.conf` is the default configuration file and includes configuration files from the folder `/ets/strongswan.d`.

The content of `/etc/strongswan.d/pki.conf`:
```
pki {

    # Plugins to load in the pki tool.
    # load =

    scep {

        # Source IP address to bind for HTTP operations.
        # http_bind =

        # Timeout for HTTP operations.
        # http_timeout = 30s

        # Some SCEP servers (e.g. openxpki) are incorrectly doing certificate
        # renewal via messageType PKCSReq (19) instead of RenewalReq (17).
        renewal_via_pkcs_req = yes

    }

}
```

The content of `/etc/strongswan.conf`:
```
# strongswan.conf - strongSwan configuration file
#
# Refer to the strongswan.conf(5) manpage for details
#
# Configuration changes should be made in the included files

charon {
	load_modular = yes
	plugins {
		include strongswan.d/charon/*.conf
	}
}

include strongswan.d/*.conf
```

---

I just would like to precise in the command line given in description `newscep.key` and `scep.key` are RSA 2048 bits private keys, PEM formatted, generated with `openssl genrsa` command.
I know from the [strongSwan documentation](https://docs.strongswan.org/docs/5.9/pki/pkiScep.html#_options) that `--in` should be:
> Public key or PKCS#10 certificate request file to issue.

But if I provide a public key or a CSR I got an error message from strongswan "building CRED_PRIVATE_KEY - RSA failed, tried 7 builders - parsing private key failed", so I concluded that I have to provide a private key through the `--in` argument.

**Reply by tobiasbrunner** [reply]
The config looks OK. So with this, do you get the same error on the server?

> But if I provide a public key or a CSR I got an error message from strongswan "building CRED_PRIVATE_KEY - RSA failed, tried 7 builders - parsing private key failed", so I concluded that I have to provide a private key through the `--in` argument.

Yes, the documentation is incorrect, only private keys are supported for `--in` (the man page and `--help` output have it correctly).

**Reply by FelixG1316** [reply]
> The config looks OK. So with this, do you get the same error on the server?

Indeed, I got the same result (HTTP status 400), but not the same error from the server. On the server, I can find in `/var/log/openxpki/scep.log` the following messages:
```
2023/05/11 08:20:03 INF SCEP handler initialized [pid=70|ep=[undef]]
2023/05/11 08:20:03 WAR Unable to detect script name - please check the docs [pid=70|ep=[undef]]
2023/05/11 08:20:05 WAR Unable to detect script name - please check the docs [pid=70|ep=]
2023/05/11 08:20:05 WAR Client error / malformed request badRequest [pid=70|ep=]
2023/05/11 08:20:05 INF Disconnect client [pid=70|ep=]
```

There is not any error, but it still declares the client's request "Bad Request", so right now I don't really know if the issue comes from the openxpki server or still from the strongswan client 😢

**Reply by tobiasbrunner** [reply]
Yeah, difficult to say. Can you provide the commands you used, so I can try to reproduce this here.

**Reply by FelixG1316** [reply]
### Server side

To install and configure the PKI, I used following commands:
```
$ git clone https://github.com/openxpki/openxpki-docker.git
$ cd openxpki-docker/
$ git clone https://github.com/openxpki/openxpki-config.git --single-branch --branch=community
$ cp contrib/wait_on_init.yaml openxpki-config/config.d/system/local.yaml
$ docker-compose up
```

Then I waited for the following messages appeared:
```
openxpki-server_1  | Setting gid to "102" 
openxpki-server_1  | Setting uid to "101" 
```

Then I entered:
```
$ docker exec -it openxpki_openxpki-server_1 /bin/bash /etc/openxpki/contrib/sampleconfig.sh
```

> From [openxpki-docker/README.md](https://github.com/openxpki/openxpki-docker/blob/master/README.md)

### Client side

```
$ openssl genrsa -out scep.key 2048
$ pki --scepca --url http://localhost:8080/scep --outform pem --raout racert --caout cacert
$ pki --scep --url http://localhost:8080/scep --outform pem --cacert cacert.pem --cacert-sig cacert-1.pem --cacert-enc racert.pem --in scep.key --dn "CN=my_hostname" --interval 10 --maxpolltime 120 > scep.crt
```

Then I approved the SCEP request from the openxpki WebUI with the test account "rob", and I got my certificate `scep.crt`.

Then, to renew my certificate, I entered:
```
$ openssl genrsa -out newscep.key 2048
$ pki --scep --url http://localhost:8080/scep --cacert ca.pem --cacert-sig ca-1.pem --cacert-enc ra.pem --in newscep.key --dn "CN=my_hostname" --cert scep.crt --key scep.key --outform pem
```

**Reply by tobiasbrunner** [reply, marked as answer]
Thanks a lot. Just for the record, the last command should be as follows to match the CA cert names of the previous command:
```
$ pki --scep --url http://localhost:8080/scep --cacert cacert.pem --cacert-sig cacert-1.pem --cacert-enc racert.pem --in newscep.key --dn "CN=my_hostname" --cert scep.crt --key scep.key --outform pem
```
However, as you already reported, this still results in an error. Increasing the log level shows that `WAR Client error / malformed request badRequest` is caused by an internal `I18N_OPENXPKI_UI_ENROLLMENT_ERROR_SIGNER_NOT_AUTHORIZED` error.

The reason for that is that the subject DN during renewal [has to match the full subject DN of the previous certificate exactly](https://openxpki.readthedocs.io/en/stable/reference/configuration/workflows/enroll.html#renewal). Since OpenXPKI (with the test config used here) adds several `DC` RDNs to the issued certificates, the full command should be like this (you might also have to create a new private key for it to work if you tried unsuccessfully before):

```
pki --scep --url http://localhost:8080/scep --cacert cacert.pem --cacert-sig cacert-1.pem --cacert-enc racert.pem --in newscep.key --dn "DC=org, DC=OpenXPKI, DC=Test Deployment, CN=my_hostname" --cert scep.crt --key scep.key --outform pem
```
To avoid this issue, we could just make `--dn` optional when `--cert` is passed and extract it from there (see the _1689-pki-scep-renew_ branch).

**Reply by FelixG1316** [reply]
Oh, indeed I missed that, and it works with the fixed command you have provided ! 

Thank you very much ! 🙏

**Comment by thrdgeek** [comment]
I realize this is an old thread and my question is this.  What version of strongswan do you have installed that includes the --scep option.
This is my version and it doesn't include that.
ipsec pki --help
strongSwan 5.9.5 PKI tool
loaded plugins: aesni aes rc2 sha2 sha1 md5 mgf1 random x509 revocation pubkey pkcs1 pkcs7 pkcs8 pkcs12 dnskey sshkey pem openssl gmp hmac drbg
usage:
  pki --verify  (-v)  verify a certificate using the CA certificate
  pki --signcrl (-c)  issue a CRL using a CA certificate and key
  pki --self    (-s)  create a self signed certificate
  pki --req     (-r)  create a PKCS#10 certificate request
  pki --pub     (-p)  extract the public key from a private key/certificate
  pki --print   (-a)  print a credential in a human readable form
  pki --pkcs7   (-7)  PKCS#7 wrap/unwrap functions
  pki --pkcs12  (-u)  PKCS#12 functions
  pki --keyid   (-k)  calculate key identifiers of a key/certificate
  pki --issue   (-i)  issue a certificate using a CA certificate and key
  pki --gen     (-g)  generate a new private key
  pki --dn      (-d)  extract the subject DN of an X.509 certificate
  pki --acert   (-z)  issue an attribute certificate
  pki --help    (-h)  show usage information

**Reply by tobiasbrunner** [reply]
You might want to check [the docs](https://docs.strongswan.org/docs/5.9/pki/pki.html) where the version numbers are mentioned for some of the newer commands, or check the [release notes](https://github.com/strongswan/strongswan/releases) (can be searched).

**Comment by thrdgeek** [comment]
Thank you.  I have read the same link you have provided.  It covers version 5.9 and I have 5.9.5 installed.  It appears the --scep option has been removed.

**Reply by tobiasbrunner** [reply]
You might want to try reading the page.
